src/filters/matches.rs

```diff
@@ -0,0 +1,159 @@
+crate::include_proto!("quilkin.extensions.filters.matches.v1alpha1");
+
+use self::quilkin::extensions::filters::matches::v1alpha1::Matches as ProtoConfig;
+use crate::filters::{prelude::*, registry::FilterRegistry};
+
+pub const NAME: &str = "quilkin.extensions.filters.matches.v1alpha1.Matches";
+
+/// Creates a new factory for generating capture filters.
+pub fn factory() -> DynFilterFactory {
+    Box::from(MatchesFactory::new())
+}
+
+struct Matches {
+    config: Config,
+    metrics_registry: prometheus::Registry,
+    filter_registry: FilterRegistry,
+}
+
+impl Matches {
+    fn new(
+        config: Config,
+        filter_registry: FilterRegistry,
+        metrics_registry: prometheus::Registry,
+    ) -> Self {
+        Self {
+            config,
+            filter_registry,
+            metrics_registry,
+        }
+    }
+}
+
+/// The actual behaviour of the matches filter, the only thing that changes
+/// between read and write is which config and direction is called.
+macro_rules! matches_filter {
+    ($this:ident, $context:ident, $direction:ident, $config:ident) => {{
+        let ctx = $context;
+        let config = match $this.config.$config.as_ref() {
+            Some(value) => value,
+            None => return Some(ctx.into()),
+        };
+
+        for branch in &config.branches {
+            if ctx.metadata.get(&config.metadata_key) == Some(&branch.value) {
+                let args = CreateFilterArgs::fixed(
+                    $this.filter_registry.clone(),
+                    $this.metrics_registry.clone(),
+                    branch.config.as_ref(),
+                )
+                .with_metrics_registry($this.metrics_registry.clone());
+
+                let instance = $this.filter_registry.get(&branch.filter, args).ok()?;
```

Comment:
This looks to be creating a new filter on each packet which wouldn't be the behavior we want? Rather Match would want to e.g instantiate the filters for all its arms once, when it itself is instantiated and that filter can keep state as usual?

Reply:
+1 on this one - I figure it would make sense for Matches to container a `FilterChain`, which could be created on `new` and stored. Execution of `read` and `write` on `Match` should be pretty straight forward at that point I think.